File: assets.js
require("dotenv").config();
const { WebflowClient } = require("webflow-api");
const crypto = require("crypto");
const axios = require("axios");
const FormData = require("form-data");

const webflow = new WebflowClient({
  accessToken: process.env.WEBFLOW_API_TOKEN,
});

// Organize folders and assets
const assetDetails = [
  {
    folderName: "Universal Assets",
    assets: ["https://images.unsplash.com/photo-1451187580459-43490279c0fa"],
  },
  {
    folderName: "English Assets",
    assets: [
      "https://images.unsplash.com/photo-1543109740-4bdb38fda756",
      "https://images.unsplash.com/photo-1665410620550-c54105af7d0c",
      "https://images.unsplash.com/photo-1526129318478-62ed807ebdf9",
    ],
  },
  {
    folderName: "French Assets",
    assets: [
      "https://images.unsplash.com/photo-1454386608169-1c3b4edc1df8",
      "https://images.unsplash.com/photo-1500039436846-25ae2f11882e",
      "https://images.unsplash.com/photo-1528717663417-3742fee05a29",
    ],
  },
];

// Function to create an asset folder
async function createAssetFolder(siteId, folderName) {
  try {
    // Check if the folder already exists
    const existingFolders = await webflow.assets.listFolders(siteId);
    const existingFolder = existingFolders.assetFolders.find(
      (folder) => folder.displayName === folderName
    );
    if (existingFolder) {
      console.log(
        `Folder '${folderName}' already exists with ID: ${existingFolder.id}`
      );
      return existingFolder.id;
    }

    // Create the folder if it does not exist
    console.log(`Creating folder: ${folderName}`);
    const response = await webflow.assets.createFolder(siteId, {
      displayName: folderName,
    });
    console.log(`Folder '${folderName}' created with ID: ${response.id}`);
    return response.id; // Return folder ID for further use
  } catch (error) {
    console.error(
      `Error creating or retrieving asset folder '${folderName}':`,
      error
    );
  }
}

// Function to hash file data from URL
async function getFileHashFromUrl(assetUrl) {
  // Create a promise to handle asynchronous hashing of the file data
  return new Promise(async (resolve, reject) => {
    try {
      // Fetch the file as a stream
      const response = await axios.get(assetUrl, { responseType: "stream" });

      // Initialize SHA-256 hash
      const hash = crypto.createHash("sha256");

      // Update the hash with each chunk of data received from the stream
      response.data.on("data", (data) => hash.update(data));

      // Finalize the hash calculation and resolve the promise with the hash value
      response.data.on("end", () => resolve(hash.digest("hex")));
      response.data.on("error", reject);
    } catch (error) {
      reject(error);
    }
  });
}

// Function to upload an asset to Webflow via S3
async function uploadAsset(siteId, folderId, assetUrl) {
  try {
    // Generate the file hash for validation
    const fileHash = await getFileHashFromUrl(assetUrl);
    const fileName = assetUrl.split("/").pop();

    // Step 1: Initialize the upload
    const uploadInit = await webflow.assets.create(siteId, {
      parentFolder: folderId,
      fileName: fileName + `.jpeg`,
      fileHash: fileHash,
    });
    const { uploadUrl, uploadDetails } = uploadInit;

    // Create form data for S3 upload
    const form = new FormData();

    // Append all required fields to the form
    form.append("acl", uploadDetails.acl);
    form.append("bucket", uploadDetails.bucket);
    form.append("X-Amz-Algorithm", uploadDetails.xAmzAlgorithm);
    form.append("X-Amz-Credential", uploadDetails.xAmzCredential);
    form.append("X-Amz-Date", uploadDetails.xAmzDate);
    form.append("key", uploadDetails.key);
    form.append("Policy", uploadDetails.policy);
    form.append("X-Amz-Signature", uploadDetails.xAmzSignature);
    form.append("success_action_status", uploadDetails.successActionStatus);
    form.append("Content-Type", uploadDetails.contentType);
    form.append("Cache-Control", uploadDetails.cacheControl);

    // Append the file to be uploaded
    const response = await axios.get(assetUrl, { responseType: "stream" });
    form.append("file", response.data, {
      filename: fileName,
      contentType: uploadDetails.contentType,
    });
    console.log(response);

    // Step 2: Upload to the provided S3 URL
    const uploadResponse = await axios.post(uploadUrl, form, {
      headers: {
        ...form.getHeaders(),
      },
    });

    if (uploadResponse.status === 201) {
      console.log(`Successfully uploaded ${fileName} to Webflow.`);
    } else {
      console.error(
        `Failed to upload ${fileName}. Response status: ${uploadResponse.status}`
      );
    }
  } catch (error) {
    console.error(`Error uploading asset from ${assetUrl}:`, error);
  }
}

// Main function to execute the upload
(async () => {
  const siteId = process.env.SITE_ID; // Replace with your actual site ID

  for (const { folderName, assets } of assetDetails) {
    const folderId = await createAssetFolder(siteId, folderName);
    if (folderId) {
      for (const assetUrl of assets) {
        await uploadAsset(siteId, folderId, assetUrl);
      }
    }
  }
})();

File: assets.py
import os
import hashlib
import requests
from webflow.client import Webflow
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Initialize Webflow client
webflow = Webflow(access_token=os.getenv("WEBFLOW_API_TOKEN"))

# Asset details
asset_details = [
    {
        "folderName": "Universal Assets",
        "assets": ["https://images.unsplash.com/photo-1451187580459-43490279c0fa"],
    },
    {
        "folderName": "English Assets",
        "assets": [
            "https://images.unsplash.com/photo-1543109740-4bdb38fda756",
            "https://images.unsplash.com/photo-1665410620550-c54105af7d0c",
            "https://images.unsplash.com/photo-1526129318478-62ed807ebdf9",
        ],
    },
    {
        "folderName": "French Assets",
        "assets": [
            "https://images.unsplash.com/photo-1454386608169-1c3b4edc1df8",
            "https://images.unsplash.com/photo-1500039436846-25ae2f11882e",
            "https://images.unsplash.com/photo-1528717663417-3742fee05a29",
        ],
    },
]

# Function to create an asset folder
def create_asset_folder(site_id, folder_name):
    existing_folders = webflow.assets.list_folders(site_id)
    # Use attribute-based access for 'display_name'
    existing_folder = next((folder for folder in existing_folders.asset_folders if folder.display_name == folder_name), None)
    if existing_folder:
        print(f"Folder '{folder_name}' already exists with ID: {existing_folder.id}")
        return existing_folder.id

    # Creating folder if not exists
    print(f"Creating folder: {folder_name}")
    response = webflow.assets.create_folder(site_id=site_id, displayName=folder_name)
    print(f"Folder '{folder_name}' created with ID: {response.id}")
    return response.id

# Function to hash file data from URL
def get_file_hash_from_url(asset_url):
    response = requests.get(asset_url, stream=True)
    hash_sha256 = hashlib.sha256()
    for chunk in response.iter_content(8192):
        hash_sha256.update(chunk)
    return hash_sha256.hexdigest()

# Function to upload an asset to Webflow via S3
def upload_asset(site_id, folder_id, asset_url):
    try:
        file_hash = get_file_hash_from_url(asset_url)
        file_name = asset_url.split("/")[-1]

        # Initiating the upload
        upload_init = webflow.assets.create(
            site_id=site_id, 
            parent_folder=folder_id,
            file_name=file_name + ".jpeg",
            file_hash=file_hash,
        )

        # Access upload details via attributes
        upload_url = upload_init.upload_url
        upload_details = upload_init.upload_details

        # Prepare the form data
        form_data = {
            "acl": upload_details.acl,
            "bucket": upload_details.bucket,
            "X-Amz-Algorithm": upload_details.x_amz_algorithm,
            "X-Amz-Credential": upload_details.x_amz_credential,
            "X-Amz-Date": upload_details.x_amz_date,
            "key": upload_details.key,
            "Policy": upload_details.policy,
            "X-Amz-Signature": upload_details.x_amz_signature,
            "success_action_status": upload_details.success_action_status,
            "Content-Type": upload_details.content_type,
            "Cache-Control": upload_details.cache_control,
        }

        # File to upload
        response = requests.get(asset_url, stream=True)
        files = {"file": (file_name, response.raw, upload_details.content_type)}

        # Upload to S3
        upload_response = requests.post(upload_url, data=form_data, files=files)
        if upload_response.status_code == 201:
            print(f"Successfully uploaded {file_name} to Webflow.")
        else:
            print(f"Failed to upload {file_name}. Status: {upload_response.status_code}")
    except Exception as e:
        print(f"Error uploading asset from {asset_url}: {e}")

# Main function to execute folder creation and asset upload
def main():
    site_id = os.getenv("SITE_ID")
    for folder in asset_details:
        folder_id = create_asset_folder(site_id, folder["folderName"])
        if folder_id:
            for asset_url in folder["assets"]:
                upload_asset(site_id, folder_id, asset_url)

if __name__ == "__main__":
    main()
